presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoinsUtils.java

```diff
@@ -0,0 +1,183 @@
+     * In order not to generate the inverse of any set, we always include the 0th
+     * node in our sets.
+     *
+     * @param totalNodes
+     * @return A set of sets each of which defines a partitioning of totalNodes
+     */
+    public static Set<Set<Integer>> generatePartitions(int totalNodes)
+    {
+        return generatePartitions(ImmutableSet.of(0), totalNodes, 1);
+    }
+
+    private static Set<Set<Integer>> generatePartitions(Set<Integer> existingSet, int totalNodes, int index)
+    {
+        ImmutableSet.Builder<Set<Integer>> sets = ImmutableSet.builder();
+        if (index < totalNodes && existingSet.size() < totalNodes - 1) {
+            ImmutableSet.Builder<Integer> setWithAddedNode = ImmutableSet.builder();
+            setWithAddedNode.addAll(existingSet);
+            setWithAddedNode.add(index);
+            sets.addAll(generatePartitions(setWithAddedNode.build(), totalNodes, index + 1));
+            sets.addAll(generatePartitions(existingSet, totalNodes, index + 1));
+        }
+        else {
+            sets.add(existingSet);
+        }
+        return sets.build();
+    }
+
+    public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> leftSources, PlanNodeIdAllocator idAllocator)
+    {
+        List<PlanNode> sources = joinGraph.getSources();
+        ImmutableList.Builder<PlanNode> leftNodes = ImmutableList.builder();
+        List<Symbol> leftSymbols = new ArrayList<>();
+        ImmutableList.Builder<PlanNode> rightNodes = ImmutableList.builder();
+        List<Symbol> rightSymbols = new ArrayList<>();
+        for (int i = 0; i < sources.size(); i++) {
+            PlanNode source = sources.get(i);
+            if (leftSources.contains(i)) {
+                leftNodes.add(source);
+                leftSymbols.addAll(source.getOutputSymbols());
+            }
+            else {
+                rightNodes.add(source);
+                rightSymbols.addAll(source.getOutputSymbols());
+            }
+        }
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> leftCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> rightCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> spanningCriteria = ImmutableList.builder();
+        for (JoinNode.EquiJoinClause equiJoinClause : joinGraph.getCriteria()) {
+            Set<Symbol> equiJoinClauseSymbols = ImmutableSet.of(equiJoinClause.getLeft(), equiJoinClause.getRight());
+            if (leftSymbols.containsAll(equiJoinClauseSymbols)) {
+                leftCriteria.add(equiJoinClause);
+            }
+            else if (rightSymbols.containsAll(equiJoinClauseSymbols)) {
+                rightCriteria.add(equiJoinClause);
+            }
+            else {
+                spanningCriteria.add(equiJoinClause);
+            }
+        }
+
+        ImmutableList.Builder<Expression> leftFilters = ImmutableList.builder();
```

Comment:
@rschlussel  In any case, I don't think you need to do filter or conditions splitting at all. You can simply add `FilterNode` on top of new join tree and rely on `PredicatePushdown` to do it's work. This will greatly simplify the code.